flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4480,18 +4559,39 @@ class ArrayExprLowering {
     std::string globalName = Fortran::lower::mangle::mangleArrayLiteral(x);
     fir::GlobalOp global = builder.getNamedGlobal(globalName);
     if (!global) {
-      global = builder.createGlobalConstant(
-          loc, arrTy, globalName,
-          [&](fir::FirOpBuilder &builder) {
-            Fortran::lower::StatementContext stmtCtx;
-            fir::ExtendedValue result =
-                Fortran::lower::createSomeInitializerExpression(
-                    loc, converter, toEvExpr(x), symMap, stmtCtx);
-            mlir::Value castTo =
-                builder.createConvert(loc, arrTy, fir::getBase(result));
-            builder.create<fir::HasValueOp>(loc, castTo);
-          },
-          builder.createInternalLinkage());
+      mlir::Type symTy = arrTy;
+      auto eleTy = symTy.cast<fir::SequenceType>().getEleTy();
+      /* If we have a rank-1 array of integer, real, or logical, then we can
+       * create a global array with the dense attribute.
+       *
+       * The mlir tensor type can only handle integer, real, or logical. It
+       * does not currently support nested structures which is required for
+       * complex.
+       *
+       * Also, we currently handle just rank-1 since tensor type assumes
+       * row major array ordering. We will need to reorder the dimensions
+       * in the tensor type to support Fortran's column major array ordering.
+       * How to create this tensor type is to be determined.
+       */
+      if (x.Rank() == 1 && (eleTy.dyn_cast<mlir::IntegerType>() ||
+                            eleTy.dyn_cast<mlir::FloatType>() ||
+                            eleTy.dyn_cast<fir::LogicalType>()))
+        global = Fortran::lower::createDenseGlobal(
```

Comment:
This is a `createGlobalConstant` variant, right, which has other invocations.  Does the original `createGlobalConstant` have the information needed to choose, and make the decision there?  If for some reason that won't fly, other options include creating a `createGlobalConstant` overload, or add a name hint like `createDenseGlobalConstant`.

Reply:
I don't think it is necessarily a variant of createGlobalConstant. It performs some tasks for building the the dense attribute then it calls createGlobal. I could still use your name suggestion if you still think it's useful.

Reply:
The code location is `CC genarr(const Fortran::evaluate::Constant<A> &x)` which is currently either finding an existing global constant, or calling `createGlobalConstant` to generate one.  The new routine is interposing on the existing call for a subset of cases where the existing code is correct but too slow.  I don't think other existing calls to `createGlobalConstant` encompass this new variant, but a future call might.  So it seems like the logic for choosing which function to call should be inside a `createGlobalConstant` overload so it would automatically apply to any future call cases, rather than in `genarr`.

If the arguments are (sufficiently) different than those of the two current `createGlobalConstant` overloads, maybe this case should be a third overload, which is declared and defined as a sibling of the existing overloads.

(I haven't gotten far enough into the details to form an opinion on which of these would be most appropriate.)

Reply:
There's another spot where createDenseGlobal() is called. That's in static fir::GlobalOp defineGlobal() of ConvertVariable.cpp. This is for explicit global declarations which are not constant arrays, but they have initialization.   Calling it createDenseGlobalConstant() may not make much sense in that case.

Reply:
Ok, code evolution in action ..  Thanks for checking; sorry for the noise.